### PR TITLE
ADD: Eq, Ord, and Show instances for JSDate

### PR DESCRIPTION
**Pull request by eric-corumdigital**

### CODE REVIEW

**Comment by eric-corumdigital** [comment]
Resolves https://github.com/purescript-contrib/purescript-js-date/issues/14

File: src/Data/JSDate.purs [hunk 1]
```diff
@@ -263,3 +272,6 @@ toTimeString dt = runFn2 dateMethod "toTimeString" dt
 -- | Returns the date as a string using the UTC timezone.
 toUTCString :: JSDate -> String
 toUTCString dt = runFn2 dateMethod "toUTCString" dt
+
+-- | Returns the date at a number of milliseconds since 1970-01-01 00:00:00 UTC.
+foreign import fromTime :: Number -> JSDate
```
Comment (LiamGoodacre) [inline src/Data/JSDate.purs:278]: `fromTime` isn't exported or used?

**Comment by eric-corumdigital** [comment]
@LiamGoodacre I have not the slightest idea of what you are talking about.

File: src/Data/JSDate.purs (hunk 1 above)
Comment (garyb) [reply src/Data/JSDate.purs:278]: It already exists as `fromInstant` too, I'd prefer we use that instead.

Comment (eric-corumdigital) [reply src/Data/JSDate.purs:278]: `fromTime` copes with `NaN` and out of bound `Number`s in the native way, which is not something that makes much sense with `Instant`.

**Comment by garyb** [comment]
`Generic` instances aren't suitable for `JSDate` - `Generic` can only accurately represent values that are PS sum or product types.

**Comment by eric-corumdigital** [comment]
I agree that this is not an obvious use of Generic but I do not know what the problem is. My understanding of Generic is that it supports encoding and decoding from types constructed from select constructors such as `Sum`, `Argument`, and so forth. The choice of these type constructors is clearly to mimic the grammar of ADT definitions so use in that case becomes obvious.

The ADT JSDate is analogous to is:

```
newtype JSDate = JSDate Number
```

Hence my choice for the Generic representation. `Number` is an accurate representaton of `JSDate` because `fromTime <<< getTime = id`.

**Comment by garyb** [comment]
Well, as an example, `genericShow` is incorrect with this instance. `Generic` is not meant to be used for foreign types.

**Comment by eric-corumdigital** [comment]
I am not invested in the intentions of Generic. I am just looking at what Generic actually is.

There is no reason to expect `show = genericShow` but lets say it is expected that, like `show`, `genericShow` should be valid PureScript code to reconstruct the value. In this case the answer is simply to change the constructor name from `"JSDate"` to `"fromTime"`.

```
> genericShow (fromTime 0.0)
"(fromTime 0.0)"
```

I will submit that revision.

**Comment by garyb** [comment]
To be completely upfront (and well, a more than a little stubborn) I'm probably not going to merge this as long as there's a `Generic` instance.

Can I ask why the instance is even useful for this type, since you've implemented the other instances that are usually generic derived anyway?

**Comment by eric-corumdigital** [comment]
No I am glad you are making the discussion straight-forward! Okay, I have reverted the addition of Generics.

File: src/Data/JSDate.purs (hunk 1 above)
Comment (garyb) [reply src/Data/JSDate.purs:278]: Can you add the `fromTime` export? Then I'll finally get this merged in and released.

Comment (garyb) [reply src/Data/JSDate.purs:278]: I guess it disappeared again in the convoluted reversion process :wink:

**Pull request merged**